### PR TITLE
test #1148 Adding importance mask Compounding mode

### PR DESCRIPTION
**Pull request by dzenanz**

### CODE REVIEW

File: src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx [hunk 1]
```diff
@@ -212,6 +226,18 @@ vtkImageData* vtkPlusPasteSliceIntoVolume::GetAccumulationBuffer()
 }
 
 //----------------------------------------------------------------------------
+vtkImageData* vtkPlusPasteSliceIntoVolume::GetImportanceMask()
+{
+  return this->ImportanceMask;
+}
+
+//----------------------------------------------------------------------------
+void vtkPlusPasteSliceIntoVolume::SetImportanceMask(vtkImageData* importanceMask)
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx]: If there's no functionality to the accessors, could you use the vtkSet/Get macros in the header?

File: src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx [hunk 2]
```diff
@@ -277,6 +303,43 @@ PlusStatus vtkPlusPasteSliceIntoVolume::ResetOutput()
                          outData->GetScalarSize()*outData->GetNumberOfScalarComponents() ) );
   }
 
+  //if (this->GetCompoundingMode() == IMPORTANCE_MASK_COMPOUNDING_MODE)
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx]: If the code is not needed, could you remove?

File: src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx [hunk 3]
```diff
@@ -299,11 +362,21 @@ PlusStatus vtkPlusPasteSliceIntoVolume::InsertSlice( vtkImageData* image, vtkMat
     return PLUS_FAIL;
   }
 
+  //vtkSmartPointer<vtkBMPWriter> writer = vtkSmartPointer<vtkBMPWriter>::New();
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx]: Same (unneeded code)

File: src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx [hunk 4]
```diff
@@ -733,6 +823,8 @@ const char* vtkPlusPasteSliceIntoVolume::GetCompoundingModeAsString( Compounding
     return "LATEST";
   case MEAN_COMPOUNDING_MODE:
     return "MEAN";
+  case IMPORTANCE_MASK_COMPOUNDING_MODE:
+	  return "IMPORTANCEMASK";
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx]: The indentation is not the same. Style guildelines are available here
[https://app.assembla.com/spaces/plus/wiki/Developers_guide#coding_conventions](https://app.assembla.com/spaces/plus/wiki/Developers_guide#coding_conventions)

File: src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolumeHelperCommon.h [hunk 5]
```diff
@@ -112,6 +112,8 @@ struct vtkPlusPasteSliceIntoVolumeInsertSliceParams
   vtkImageData* outData;            // the output volume
   void* outPtr;                     // scalar pointer to the output volume over the output extent
   unsigned short* accPtr;           // scalar pointer to the accumulation buffer over the output extent
+  vtkImageData* imData;             // importance mask
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolumeHelperCommon.h]: I know the other variables aren't very good, but would you mind renaming these to be more descriptive. If you're feeling generous, you could rename some others!

File: src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.cxx [hunk 6]
```diff
@@ -883,3 +902,55 @@ void vtkPlusVolumeReconstructor::SetFanAnglesAutoDetectFilterRadiusPixel(int rad
 {
   this->FanAngleDetector->SetFilterRadiusPixel(radiusPixel);
 }
+
+//----------------------------------------------------------------------------
+void vtkPlusVolumeReconstructor::SetImportanceMaskFilename(std::string filename)
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.cxx]: const std::string& please!

File: src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.cxx [hunk 7]
```diff
@@ -883,3 +902,55 @@ void vtkPlusVolumeReconstructor::SetFanAnglesAutoDetectFilterRadiusPixel(int rad
...
+{
+  if (this->ImportanceMaskFilename != filename)
+  {
+    this->ImportanceMaskFilename = filename;
+    
+    bool useVTK = true;
+    if (useVTK) //do not drag in ITK as a dependency
+    {
+      vtkSmartPointer<vtkPNGReader> reader = vtkSmartPointer<vtkPNGReader>::New();
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.cxx]: Is the CMake currently set up to link to ITK? If so, go for it. If not, maybe remove the ITK section.

File: src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.h [hunk 8]
```diff
@@ -191,6 +191,9 @@ class vtkPlusVolumeReconstructionExport vtkPlusVolumeReconstructor : public vtkI
   void SetPixelRejectionThreshold(double threshold);
   double GetPixelRejectionThreshold();
 
+  std::string GetImportanceMaskFilename();
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.h]: Could you have this function return a const& and make the function const

File: src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.h [hunk 9]
```diff
@@ -191,6 +191,9 @@ class vtkPlusVolumeReconstructionExport vtkPlusVolumeReconstructor : public vtkI
...
+  void SetImportanceMaskFilename(std::string filename);
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.h]: const std::string&

**Comment by adamrankin** [comment]
Thanks for the addition! I've done some basic code checks.

We don't have a build system setup for pull requests yet. Have you performed fairly thorough testing? If so, looks good to me after minor fixes!

File: src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.cxx (hunk 7 above)
Comment (dzenanz) [inline src/PlusVolumeReconstruction/vtkPlusVolumeReconstructor.cxx]: It is set up to link to ITK. Should I remove the VTK variant then?

**Comment by dzenanz** [comment]
Should I remove VTK way of reading the importance mask?

P.S. Some lines were marked as modified even though they aren't. Probably because of copying code from old SVN repository (which I have compiled) and new Git repository which I currently use only for this pull request.

File: src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx [hunk 10]
```diff
@@ -823,8 +765,8 @@ const char* vtkPlusPasteSliceIntoVolume::GetCompoundingModeAsString( Compounding
     return "LATEST";
   case MEAN_COMPOUNDING_MODE:
     return "MEAN";
-  case IMPORTANCE_MASK_COMPOUNDING_MODE:
-	  return "IMPORTANCEMASK";
+  case IMPORTANCE_MASK_COMPOUNDING_MODE:
+	return "IMPORTANCEMASK";
```
Comment (adamrankin) [inline src/PlusVolumeReconstruction/vtkPlusPasteSliceIntoVolume.cxx]: I think this is a tab instead of spaces? I will update our hook to check for this.

**Comment by adamrankin** [comment]
Yes, if ITK is already linked, feel free to remove VTK code.

Re: old code, thanks for cleaning it up!

@lassoan Looks good to me, I'll let you do the final inspection/acceptance.

**Comment by lassoan** [comment]
VTK is nice because it handles all voxel types. Maybe leave in the VTK reader and if it fails then give the ITK reader a chance to read the image.

**Comment by dzenanz** [comment]
The logic requires unsigned char, e.g. pointers are cast to uchar, so there is no benefit to VTK's voxel type polymorphism there. I just kept ITK reader. Trying to first read with VTK produced a nasty error in the log saying that it is not a valid PNG image when I supplied .nrrd version.

**Comment by dzenanz** [comment]
I tested this code fairly thoroughly (all 6 combinations of linear/NN and no/partial/full optimization). However I was only using VS2015. I don't think I used any fancy language features, so it should compile everywhere. But we won't know that until we test :smile: 

We can merge, as far as I am concerned.

**Comment by dzenanz** [comment]
@lassoan @adamrankin Can you merge this before it gets out of date again?

**Comment by adamrankin** [comment]
I'd like @lassoan to approve this, because he is the most familiar with the volume reconstruction code.

**Comment by adamrankin** [comment]
But as it stands it passes my inspection.

**Comment by dzenanz** [comment]
Great!

**Comment by lassoan** [comment]
@dzenanz, Integrating this branch caused a lot of build errors - see http://crunch.cs.queensu.ca/CDash/viewBuildError.php?buildid=65803 -- Seems to be some ITK link errors. Could you have a look and if you figured out a solution send a pull request? If you don't have time then let us know. Thanks!

**Comment by dzenanz** [comment]
Interestingly enough, it happens on VS14 Win7 too. At least that is surprising as I developed on that platform.

I was just about to go home, but I will stay and try to add explicit dependence on ITK of VolumeReconstructor.

**Pull request merged**